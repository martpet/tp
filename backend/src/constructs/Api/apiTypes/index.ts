export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
