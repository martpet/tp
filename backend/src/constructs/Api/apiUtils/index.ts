export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
