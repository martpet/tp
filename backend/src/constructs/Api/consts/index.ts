export * from './authorizationHeader';
export * from './defaultCachePolicyProps';
export * from './defaultSecurityHeadersBehavior';
export * from './lambdaEdgeViewerEvent';
