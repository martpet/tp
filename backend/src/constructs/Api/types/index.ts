export * from './EndpointCookies';
export * from './EndpointEnvVars';
export * from './EndpointHeaders';
export * from './EndpointsQueryStrings';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
