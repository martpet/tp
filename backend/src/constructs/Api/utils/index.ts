export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './handlerWrapper';
export * from './itResolvesWithErrorResponse';
export * from './itResolvesWithLambdaEdgeErrorResponse';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
