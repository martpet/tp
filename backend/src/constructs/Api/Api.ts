import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/authEdgeFunction/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'HttpApi');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'Alias', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
