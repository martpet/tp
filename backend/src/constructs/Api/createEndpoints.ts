import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import {
  CachePolicy,
  Distribution,
  OriginRequestCookieBehavior,
  OriginRequestHeaderBehavior,
  OriginRequestPolicy,
  OriginRequestPolicyProps,
  OriginRequestQueryStringBehavior,
} from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { Writable } from 'type-fest';

import { capitalize, createNodejsFunction } from '~/utils';

import { EndpointsOptions, MethodOptions } from './apiTypes';

type CreateEndpointProps = {
  scope: Construct;
  distribution: Distribution;
  api: HttpApi;
  origin: HttpOrigin;
  options: EndpointsOptions;
};

export const createEndpoints = ({
  scope,
  distribution,
  api,
  origin,
  options,
}: CreateEndpointProps) => {
  Object.entries(options).forEach(
    ([path, { methods, cookies, headers, queryStrings }]) => {
      addDistroBehavior({
        scope,
        distribution,
        origin,
        path,
        headers,
        cookies,
        queryStrings,
      });

      Object.entries(methods).forEach(([method, methodOptions]) => {
        addApiRoute({ scope, api, path, method, methodOptions });
      });
    }
  );
};

type AddDistroBehaviorProps = {
  scope: Construct;
  distribution: Distribution;
  origin: HttpOrigin;
  path: string;
  headers?: string[];
  cookies?: string[];
  queryStrings?: Readonly<string[]>;
};

function addDistroBehavior({
  scope,
  distribution,
  origin,
  path,
  cookies,
  headers,
  queryStrings,
}: AddDistroBehaviorProps) {
  const pathName = path.replace('/', '');
  const policyProps: Writable<OriginRequestPolicyProps> = {};

  if (headers) {
    policyProps.headerBehavior = OriginRequestHeaderBehavior.allowList(...headers);
  }

  if (cookies) {
    policyProps.cookieBehavior = OriginRequestCookieBehavior.allowList(...cookies);
  }

  if (queryStrings) {
    policyProps.queryStringBehavior = OriginRequestQueryStringBehavior.allowList(
      ...queryStrings
    );
  }

  distribution.addBehavior(path, origin, {
    cachePolicy: CachePolicy.CACHING_DISABLED,
    originRequestPolicy: new OriginRequestPolicy(scope, pathName, policyProps),
  });
}

type AddApiRouteProps = {
  scope: Construct;
  api: HttpApi;
  path: string;
  method: string;
  methodOptions: MethodOptions<unknown>;
};

function addApiRoute({ scope, api, path, method, methodOptions }: AddApiRouteProps) {
  const pathName = path.replace('/', '');
  const methodName = method.toLowerCase();
  const fileName = `${methodName}-${pathName}`;
  const id = `${capitalize(methodName)}${capitalize(fileName)}`;
  const { environment, cb } = methodOptions;
  const handlerProps: Writable<NodejsFunctionProps> = {
    entry: `${__dirname}/handlers/${pathName}/${fileName}/${fileName}.ts`,
    environment,
  };

  const handler = createNodejsFunction(scope, `${id}Handler`, handlerProps);

  if (cb) {
    cb(handler);
  }

  api.addRoutes({
    path,
    methods: [method as HttpMethod],
    integration: new HttpLambdaIntegration(`${id}Integration`, handler),
  });
}
