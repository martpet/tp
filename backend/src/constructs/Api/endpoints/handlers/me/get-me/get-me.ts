import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { EventHeaders } from '~/constructs/Api/types';
import { errorResponse, getIdTokenPayload } from '~/constructs/Api/utils';

export const handler: APIGatewayProxyHandlerV2 = async ({ headers }) => {
  const { authorization: idToken } = headers as EventHeaders<'/me'>;

  if (!idToken) {
    return errorResponse('RyFuj-_6Qo');
  }

  const { givenName } = getIdTokenPayload(idToken);

  return givenName;
};
