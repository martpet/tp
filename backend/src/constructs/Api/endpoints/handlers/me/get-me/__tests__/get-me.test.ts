import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { getIdTokenPayload, itResolvesWithErrorResponse } from '~/constructs/Api/utils';
import { Me } from '~/types';

import { handler } from '../get-me';

vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');

const args = [
  {
    headers: {
      authorization: 'dummyIdToken',
    },
  },
] as unknown as Parameters<APIGatewayProxyHandlerV2<Me>>;

describe('get-me', () => {
  it('calls "getIdTokenPayload" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(getIdTokenPayload).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  describe('when "authorization" header is missing', () => {
    const argsClone = structuredClone(args);
    argsClone[0].headers.authorization = undefined;

    itResolvesWithErrorResponse(handler, argsClone);
  });
});
