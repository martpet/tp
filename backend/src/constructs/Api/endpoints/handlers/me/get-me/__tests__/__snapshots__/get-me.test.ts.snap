// Vitest Snapshot v1

exports[`get-me > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyIdToken",
  ],
]
`;

exports[`get-me > resolves with a correct value 1`] = `"dummyGivenName"`;

exports[`get-me > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "RyFuj-_6Qo",
  ],
]
`;

exports[`get-me > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;
